File: src/components/calculateAngle.js

export const calculateAngle = (a, b, c) => {
  const v1 = [b[0] - a[0], b[1] - a[1]];
  const v2 = [c[0] - b[0], c[1] - b[1]];

  const dot = v1[0] * v2[0] + v1[1] * v2[1];
  const mag1 = Math.sqrt(v1[0] * v1[0] + v1[1] * v1[1]);
  const mag2 = Math.sqrt(v2[0] * v2[0] + v2[1] * v2[1]);

  const angle = Math.acos(dot / (mag1 * mag2));
  return (angle * 180) / Math.PI;
};

File: pushup-counter-using-mediapipe/src/calculateAngle.py
# src/calculateAngle.py

import math
import numpy as np

def calculate_angle(a, b, c):
    # Calculate the angle between three points
    a = np.array(a)  # First point
    b = np.array(b)  # Midpoint
    c = np.array(c)  # End point

    # Vector AB and BC
    ab = a - b
    bc = c - b

    # Dot product and magnitudes
    dot_product = np.dot(ab, bc)
    magnitude_ab = np.linalg.norm(ab)
    magnitude_bc = np.linalg.norm(bc)

    # Angle in degrees
    angle = np.arccos(dot_product / (magnitude_ab * magnitude_bc))  # Radians
    angle = np.degrees(angle)  # Convert to degrees
    return angle
